packages/devtools_flutter/lib/src/config.dart

```diff
@@ -10,28 +13,147 @@ import 'screen.dart';
-      EmptyScreen.timeline,
-      EmptyScreen.performance,
-      EmptyScreen.memory,
-      EmptyScreen.logging,
-    ],
+  /// The routes that the app exposes.
+  final Map<String, WidgetBuilder> _routes = {
+    '/': (_) => Initializer(
+          builder: (_) => const DevToolsScaffold(
+            tabs: [
+              EmptyScreen.inspector,
+              EmptyScreen.timeline,
+              EmptyScreen.performance,
+              EmptyScreen.memory,
+              EmptyScreen.logging,
+            ],
+          ),
+        ),
+    '/connect': (_) => DevToolsScaffold.withChild(child: ConnectScreenBody()),
   };
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      theme: ThemeData.light(),
+      onGenerateRoute: _generateRoute,
+    );
+  }
+}
+
+/// Widget that requires business logic to be loaded before building its
+/// [builder].
+///
+/// See [_InitializerState.build] for the logic that determines whether the
+/// business logic is loaded.
+///
+/// Use this widget to wrap pages that require [service.serviceManager] to be
+/// connected. As we require additional services to be available, add them
+/// here.
+class Initializer extends StatefulWidget {
+  const Initializer({Key key, @required this.builder})
+      : assert(builder != null),
+        super(key: key);
+
+  /// The builder for the widget's children.
+  ///
+  /// Will only be built if [_InitializerState._loaded] is true.
+  final WidgetBuilder builder;
+
+  @override
+  _InitializerState createState() => _InitializerState();
+}
+
+class _InitializerState extends State<Initializer> {
+  final List<StreamSubscription> _subscriptions = [];
+
+  /// Checks if the [service.serviceManager] is connected.
+  ///
+  /// This is a method and not a getter to communicate that its value may
+  /// change between successive calls.
+  bool _loaded = service.serviceManager.hasConnection;
+
+  @override
+  void initState() {
+    super.initState();
+    _connectToServiceManager();
+  }
+
+  @override
+  void dispose() {
+    for (var s in _subscriptions) {
+      s.cancel();
+    }
+    super.dispose();
+  }
+
+  /// Listens to the [service.serviceManager] and rebuilds the page when the
+  /// connection status changes.
+  void _connectToServiceManager() {
+    _subscriptions.add(
+      service.serviceManager.onStateChange.listen((connected) {
+        setState(() {
+          _loaded = connected;
+        });
+      }),
+    );
+    // TODO(https://github.com/flutter/devtools/issues/1150): Check the route
+    // parameters for a VM Service URL and attempt to connect to it without
+    // going to the /connect page.
+    if (!_loaded && ModalRoute.of(context).isCurrent) {
+      // If this route is on top and the app is not loaded, then we navigate to
+      // the /connect page to get a VM Service connection for serviceManager.
+      WidgetsBinding.instance.addPostFrameCallback((_) {
+        Navigator.of(context).pushNamed('/connect');
+      });
```

Comment:
I wonder if pushing the connect screen when `loaded` is false will cause any flashy behavior where we flash to the connect screen and back to the actual screen if connection immediately follows.

Reply:
After pushing the connect screen, we will stay there.

When we try to set up connections without the connect screen with the URL parameters (see the TODO above for #1144), then we'll have to decide that we won't push a new route until after we know we can't connect.

Reply:
sgtm